main/guides/zoe/contract-upgrade.md

```diff
@@ -0,0 +1,186 @@
+  rooms = makeScalarBigMapStore('rooms', { durable: true });
+  baggage.init('rooms', rooms);
+} else {
+  // subsequent incarnation: use the object from the initial incarnation
+  rooms = baggage.get('rooms');
+}
+```
+
+The `provide` function supports a concise idiom for this find-or-create pattern:
+
+```js
+import { provide } from '@agoric/vat-data';
+
+const rooms = provide(baggage, 'rooms', () =>
+  makeScalarBigMapStore('rooms', { durable: true }),
+);
+```
+
+The `zone` API is a convenient way to manage durability. Its store methods integrate the `provide` pattern:
+
+::: details import { makeDurableZone } ...
+
+<<< @/snippets/zoe/src/02b-state-durable.js#import-zone
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#zone1
+
+::: details What happens if we don't use baggage?
+
+When the contract instance is restarted, it gets a fresh [heap](../js-programming/#vats-the-unit-of-synchrony), so [ordinary heap state](./contract-basics.html#state) does not survive upgrade. This implementation does not persist the effect of `E(publicFacet).set(2)` from the first incarnation:
+
+<<< @/snippets/zoe/src/02-state.js#heap-state{2}
+
+Also, it creates a fresh `publicFacet` object each time. So messages
+from clients that try to use the `publicFacet` from the first incarnation
+will fail to reach the `publicFacet` created in later incarnations.
+
+<<< @/snippets/zoe/src/02-state.js#fresh-export{2}
+
+:::
+
+### Kinds
+
+Use `zone.exoClass()` to define state and methods of kinds of durable objects such as `Room`:
+
+::: details import { makeDurableZone } ...
+
+<<< @/snippets/zoe/src/02b-state-durable.js#import-provide-zone{1}
+
+:::
+
+<<< @/snippets/zoe/src/02b-state-durable.js#exoclass
+
+Defining `publicFacet` as a singleton `exo` allows clients to
+continue to use it after an upgrade:
+
+<<< @/snippets/zoe/src/02b-state-durable.js#exo
+
+Now we have all the parts of an upgradable contract.
+
+::: details full contract listing
+
+<<< @/snippets/zoe/src/02b-state-durable.js#contract
+
+:::
+
+We can then upgrade it to have another method:
+
+```js
+  const makeRoom = zone.exoClass('Room', RoomI, (id) => ({ id, value: 0 }), {
+    ...
+    clear(delta) {
+      this.state.value = 0;
+    },
+  });
+```
+
+The interface guard also needs updating.
+<small>_See [@endo/patterns](https://endojs.github.io/endo/modules/_endo_patterns.html) for more on interface guards._</small>
+
+```js
+const RoomI = M.interface('Room', {
+  ...
+  clear: M.call().returns(),
+});
+```
+
+::: tip Notes
+
+- Once the state is defined by the `init` function (3rd arg), properties cannot be added or removed.
+- Values of state properties must be serializable.
+- Values of state properties are hardened on assignment.
+- You can replace the value of a state property (e.g. `state.zot = [...state.zot, 'last']`), and you can update stores (`state.players.set(1, player1)`), but you cannot do things like `state.zot.push('last')` nor `state.jot = { x: 1 }; state.jot.x = 2`.
```

Comment:
It's unfortunate that `state.jot = { x: 1 }; state.jot.x = 2` broke across a line for me. What do you think of making it a multiline expression to make it more apparent?

Reply:
added a line-break